ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java

```diff
@@ -306,9 +307,8 @@ private BlockImportResult importBlockAndState(
       final int secondsPerSlot = spec.getSecondsPerSlot(block.getSlot());
       final UInt64 timeIntoSlot =
           transaction.getTime().minus(transaction.getGenesisTime()).mod(secondsPerSlot);
```

Comment:
hmm, I don't think we're being fair here.  The idea is that if the block is received prior to the point where attestations are due it receives an extra "proposer boost" in scoring but doesn't get it if it's received after the attestations are due (thus late blocks are less likely to wind up on the canonical chain).

With this version we're rounding (or truncating) the block received time to the second but we really want to get the received time in mills which will bias them either to get or miss the boost more often.  It's not an issue for anything except Gnosis because other network's slot time divides by 3 evenly.

I suspect we're going to have to move to having `Store.time` (which is what `transaction.getTime()` is getting here) track milliseconds rather than seconds but I think that should be a separate PR. This one is fine as-is since it doesn't make things any worse (doing the comparison on seconds will bias the boost on gnosis anyway).

I'd suggest we keep `getTime()` returning seconds (maybe rename to `getTimeSeconds()`) and add a new `getTimeMillis()` method so we're very clear about when we're using seconds vs millis.  The key pieces of code for that I think will start from `BeaconChainController.onTick` which is getting the time as seconds through to `StoreTransaction.setTime`. Work backwards by finding callers of `setTime` and you'll find the path easily enough I think.

Reply:
Thank you, will open another PR for the recommended changes.